source/common/network/lc_trie.h

```diff
@@ -0,0 +1,540 @@
+
+        // count is the number of entries in the base vector that have the same bit pattern as the
+        // ip_prefixes_[new_position].
+        int count = 0;
+        while (new_position + count < first + n &&
+               static_cast<uint32_t>(extractBits<IpType, address_size>(
+                   output.prefix_, output.branch_, ip_prefixes_[new_position + count].ip_)) ==
+                   bit_pattern) {
+          ++count;
+        }
+
+        // This logic was taken from
+        // https://github.com/beevek/libkrb/blob/24a224d3ea840e2e7d2926e17d8849aefecc1101/krb/lc_trie.hpp#L396.
+        if (count == 0) {
+          if (new_position == first + n) {
+            buildRecursive(output.prefix_ + output.branch_, new_position - 1, 1,
+                           address + bit_pattern, next_free_index);
+          } else {
+            buildRecursive(output.prefix_ + output.branch_, new_position, 1, address + bit_pattern,
+                           next_free_index);
+          }
+        } else if (count == 1 &&
+                   ip_prefixes_[new_position].length_ - output.prefix_ < output.branch_) {
+          uint32_t bits = output.branch_ + output.prefix_ - ip_prefixes_[new_position].length_;
+          for (uint32_t i = bit_pattern; i < bit_pattern + (1 << bits); ++i) {
+            buildRecursive(output.prefix_ + output.branch_, new_position, 1, address + i,
+                           next_free_index);
+          }
+        } else {
+          // Recursively build sub-tries for ip_prefixes_[new_position] to
+          // ip_prefixes_[new_position+count-1].
+          buildRecursive(output.prefix_ + output.branch_, new_position, count,
+                         address + bit_pattern, next_free_index);
+        }
+        new_position += count;
+      }
+    }
+
+    /**
+     * LcNode is a uint32_t. A wrapper is provided to simplify getting/setting the branch, the skip
+     * and the address values held within the structure.
+     *
+     * The LcNode has three parts to it
+     * - Branch: the first 5 bits represent the branching factor. The branching factor is used to
+     * determine the number of descendants for the current node. The number represents a power of 2,
+     * so there can be at most 2^31 descendant nodes.
+     * - Skip: the next 7 bits represent the number of bits to skip when looking at an IP address.
+     * This value can be between 0 and 127, so IPv6 is supported.
+     * - Address: the remaining 20 bits represent an index either into the trie_ or the
+     * ip_prefixes_. If branch_ != 0, the index is for the trie_. If branch == zero, the index is
+     * for the ip_prefixes_.
+     *
+     * Note: If more than 2^19 CIDR ranges are to be stored in trie_, uint64_t should be used
+     * instead.
+     */
+    struct LcNode {
+      uint32_t branch_ : 5;
+      uint32_t skip_ : 7;
+      uint32_t address_ : 20;
+    };
+
+    // Vector that contains the CIDR ranges and tags.
```

Comment:
explain that this is kept because you need to make sure that you've got a match. I know this is based on the paper, but it might be worth it to spell that here.